bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/checksum/DigestManager.java

```diff
@@ -49,14 +49,89 @@ public abstract class DigestManager {
 
     public static final int METADATA_LENGTH = 32;
     public static final int LAC_METADATA_LENGTH = 16;
+    private static final int MAX_SUB_BUFFER_VISIT_RECURSION_DEPTH = 10;
 
     final long ledgerId;
     final boolean useV2Protocol;
     private final ByteBufAllocator allocator;
 
     abstract int getMacCodeLength();
 
-    abstract int update(int digest, ByteBuf buffer, int offset, int len);
+    abstract int internalUpdate(int digest, ByteBuf buffer, int offset, int len);
+
+    final int update(int digest, ByteBuf buffer, int offset, int len) {
+        return recursiveSubBufferVisitForDigestUpdate(digest, buffer, offset, len, 0);
+    }
+
+    private int recursiveSubBufferVisitForDigestUpdate(int digest, ByteBuf buffer, int offset, int len, int depth) {
+        if (len == 0) {
+            return digest;
+        }
+        if (depth < MAX_SUB_BUFFER_VISIT_RECURSION_DEPTH && !buffer.hasMemoryAddress() && !buffer.hasArray()) {
+            return visitWrappedBuffersAndCallInternalUpdateForEach(digest, buffer, offset, len, depth);
+        }
+        return internalUpdate(digest, buffer, offset, len);
+    }
+
+    /**
+     * This method is used to visit the wrapped buffers and call internalUpdate for each of them.
+     * CompositeByteBuf is one of the wrapped buffer types that will be visited. It can contain multiple
+     * sub buffers. The sub buffers can be wrapped buffers as well.
+     *
+     * Netty doesn't provide an API to visit the wrapped buffers, so we have to use the a hack here.
+     *
+     * @param digest current digest value
+     * @param buffer the buffer to visit
+     * @param offset the offset in the buffer
+     * @param len the length in the buffer
+     * @param depth the recursion depth of the wrapped buffer visit
+     * @return updated digest value
+     */
+    private int visitWrappedBuffersAndCallInternalUpdateForEach(int digest, ByteBuf buffer, int offset, int len,
+                                                                int depth) {
+        // hold the digest in a MutableInt so that it can be updated in the wrapped buffer visit
+        MutableInt digestRef = new MutableInt(digest);
+        ByteBuf sliced = buffer.slice(offset, len);
+        // call getBytes to trigger the wrapped buffer visit
+        sliced.getBytes(0, new DuplicatedByteBuf(Unpooled.EMPTY_BUFFER) {
```

Comment:
`DuplicatedByteBuf` is deprecated, so it could be removed in a future release of Netty.

This implementation is tricky and hard to understand without diving into the internal. After debugging step by step, I found the correctness depends on the implementation details in `UnsafeByteBufUtil`.

<img width="1018" alt="image" src="https://github.com/apache/bookkeeper/assets/18204803/808d9b8f-1b29-4a6a-99d3-b624cd41e669">

Is it possible to achieve the same goal with the following logic? 

<img width="1048" alt="image" src="https://github.com/apache/bookkeeper/assets/18204803/e65c7fd4-c847-4984-b4dd-7154005f6ee2">

Reply:
> `DuplicatedByteBuf` is deprecated, so it could be removed in a future release of Netty.

That's true. However if that would happen, the usage could be replaced. I could also replace that in this PR in further commits. I was hoping to explain the solution for achieving checksum calculations without extra copies by minimizing the changes in this PR in the first place. 

I also looked in the Netty source code while trying to find the correct "unwrapping" logic to replace the invalid solution that was introduced by #2701 . 
This is the comment I found when I knew that this doesn't exist:
https://github.com/netty/netty/blob/2c8b3fb2dea00903843a6f78ce8990b06e0bff96/handler/src/main/java/io/netty/handler/ssl/SslHandler.java#L1118-L1120

I was planning to propose a public API to Netty ByteBuf for visiting the internal buffers. This PR could be a good starting point.

> This implementation is tricky and hard to understand without diving into the internal. After debugging step by step, I found the correctness depends on the implementation details in UnsafeByteBufUtil.

I don't think that there's a way around it. There is a specific implicit contract that Netty ByteBufs follow regarding the `hasArray`, `hasMemoryAddress` and `nioBufferCount` and how that impacts what method in the destination ByteBuf the call to the source ByteBuf's `getBytes` (with `ByteBuf` dest) will be calling.

> Is it possible to achieve the same goal with the following logic?

The Netty ByteBuf contains a lot of classes that are package private. The example of `CompositeByteBuf.getBytes` will be used in the solution that this PR provides.

We have a failing test case that doesn't touch the internals. I'm open to any solution that fixes the checksum calculation bug, that the test case (the original commit works for unchanged master branch and possibly other branches, that is commit https://github.com/lhotari/bookkeeper/commit/9fb47f5dcf3848320991c7a2cf602eb325893a42). 

@BewareMyPower do you have another proposal for the direction to take in fixing this issue?

Reply:
@BewareMyPower I have address your review feedback and attempted to cover your concerns by refactoring the ByteBuf visiting solution. It no longer uses `DuplicatedByteBuf` as the base class. Instead a full custom `ByteBuf` implementation is used with `throw new UnsupportedOperationException()` bodies for all other methods than the ones that are part of the `getBytes`/`setBytes` interface contract. This will ensure that things will break if something unexpected is passed as input.